### ISSUE TITLE
autoGrantPermissions feature doesn't work if Android application has certain permissions

### ISSUE THREAD

**Issue by ilyanesterov** [appium/appium#10018]
## The problem

my simple app is writing a file to external storage and requests user permissions. I have a simple test to verify file content, however, if I set `autoGrantPermissions` to `true` in  desired capabilities, it doesn't work and fail with the following error in appium server log:
```
[ADB] Running '/Users/user/Android/platform-tools/adb' with args: ["-P",5037,"-s","025ccb49455bc8a6","shell","pm","dump","com.blah.blah.info"]
[AndroidDriver] Unable to grant permissions requested. Original error: Unable to get requested permissions
```

when I look for permissions for appium settings app it seems it can successfully parse them:

```
[debug] [ADB] Running '/Users/user/Android/platform-tools/adb' with args: ["-P",5037,"-s","025ccb49455bc8a6","shell","pm","dump","io.appium.settings"]
[debug] [ADB] Got the following command chunks to execute: pm,grant,io.appium.settings,android.permission.WRITE_SETTINGS,;,pm,grant,io.appium.settings,android.permission.ACCESS_MOCK_LOCATION,;
[debug] [ADB] Getting connected devices...
```
but not in case of my app. 

desired capabilities defined in my python test:
```python
        desired_caps = {
            'platformName': 'Android',
            'platformVersion': '6.0',
            'appPackage': 'com.blah.blah.info',
            'appActivity': '.MainActivity',
            'deviceName': '025ccb49455bc8a6',
            'autoGrantPermissions': True,
            'app': PATH(APP_PATH)
        }
```

## Environment

* Appium version (or git revision) that exhibits the issue: 1.7.2
* Last Appium version that did not exhibit the issue (if applicable):
* Desktop OS/version used to run Appium: MacOS 10.12.6 
* Node.js version (unless using Appium.app|exe): v5.5.0
* Mobile platform/version under test: Android 
* Real device or emulator/simulator: real device and emulator
* Appium CLI or Appium.app|exe: Appium CLI

## Details

please see above in the problem description section.

please see [this comment](#issuecomment-358804546) with explanation of what is happening

**Comment by ilyanesterov** [comment]
I looked through issues here and inside python client repo, but nothing really helped

**Comment by ilyanesterov** [comment]
ok. So I dig a little bit more and it seems there some issue with parsing permissions in android-driver.
My app `AndroidManifest.xml` has only following permissions:

```
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
```

and `autoGrantPermissions` doesn't work. However, if I add `<uses-permission android:name="android.permission.INTERNET" />` or `<uses-permission android:name="android.permission.INTERNET" />` might be something else (I actually looked into the source of https://github.com/appium/io.appium.settings/blob/master/app/src/main/AndroidManifest.xml because permissions seem properly parsed for that app) everything works fine and permissions are parsed and granted properly for my app.

Can someone who knows the code of, I guess android-drive point me to a place where we parse the manifest file or package permissions. I guess issue somewhere there.

**Comment by mykola-mokhnach** [comment]
the original code can be found at https://github.com/appium/appium-adb/blob/05cd2a48770ad925e5b241adee6155d7d001e234/lib/tools/adb-commands.js#L157

Your error happens when the code tries to get requested permissions list using `pm` tool, so I'd start from there: https://github.com/appium/appium-adb/blob/05cd2a48770ad925e5b241adee6155d7d001e234/lib/tools/adb-commands.js#L307

**Comment by ilyanesterov** [comment]
@mykola-mokhnach got it. That is helpful. I'll try to see what is going wrong with `getReqPermissions` in my case tomorrow.

**Comment by ilyanesterov** [comment]
@mykola-mokhnach  BTW if you have any other ideas what I can try or where I can dig in, just let me know.

**Comment by mykola-mokhnach** [comment]
I believe it's some compatibility problem with the configuration of the application under test and the `pm` tool itself. Googling should help

**Comment by ilyanesterov** [comment]
So this is an example of stdout of `pm` `dump` which doesn't work https://gist.github.com/ilyanesterov/61d3b5b0759fd05ecd74f7eb0a9062c7
And this is an example of stdout when it works:
https://gist.github.com/ilyanesterov/3d4c8ebb5cac17132edc1dad5834e5ee

I'll put pieces which interesting for us below. This is the example of nor working output:
```
      requested permissions:
        android.permission.WRITE_EXTERNAL_STORAGE
        android.permission.READ_EXTERNAL_STORAGE
      User 0:  installed=true hidden=false stopped=true notLaunched=false enabled=0
        runtime permissions:
          android.permission.READ_EXTERNAL_STORAGE: granted=true
          android.permission.WRITE_EXTERNAL_STORAGE: granted=true
```

and the example of working output
```
requested permissions:
        android.permission.INTERNET
        android.permission.WRITE_EXTERNAL_STORAGE
        android.permission.READ_EXTERNAL_STORAGE
      install permissions:
        android.permission.INTERNET: granted=true
      User 0:  installed=true hidden=false stopped=false notLaunched=false enabled=0
        gids=[3003]
        runtime permissions:
```

here is a method in `adb-commands.js` which parses the output:
```javascript
methods.getReqPermissions = async function (pkg, cmdOutput = null) {
  let stdout = cmdOutput || await this.shell(['pm', 'dump', pkg]);
  let match = new RegExp(/requested permissions:([\s\S]*?)install permissions:/g).exec(stdout);
  if (!match) {
    throw new Error('Unable to get requested permissions');
  }
  return match[0].match(/android\.permission\.\w+/g) || [];
};
```
full code  as @mykola-mokhnach  pointed out is here:
https://github.com/appium/appium-adb/blob/05cd2a48770ad925e5b241adee6155d7d001e234/lib/tools/adb-commands.js#L307

as you can see it all boils down to the way how we parse stdout of `pm dump` command. And the way we do it in regex we always assume that permissions will be between words `requested permissions:` and `install permissions:` which is not quite correct, because in some cases (as you can see by looking into not working example above, there is no closing `install permissions:` ).

So in order to fix it here is a modified regex (i'm not good in writing regular expressions, but this will fix the issue):
```
/requested permissions:([\s\S]*?)(install permissions:|User 0)/g
```

and here is a quick test I wrote using a modified getReqPermissions method:
https://gist.github.com/ilyanesterov/c3981c0368aa646501eadf3610384fa5

I'll create a pull request under appium-adb.

**Comment by mykola-mokhnach** [comment]
nicely done. We'll be happy to review the PR

**Comment by ilyanesterov** [comment]
The issue is fixed in appium-adb@5.6.1 per https://github.com/appium/appium-adb/pull/289. Closing the issue.
